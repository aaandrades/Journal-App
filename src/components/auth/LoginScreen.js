import React from "react";
import { Link } from "react-router-dom";
import { useForm } from "../../Hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import {
  initLoginAction,
  startGoogleLoginAction,
} from "../../Redux/Actions/authActions";
import { LoginState } from "../../states/InitialStates";

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const { loading } = useSelector((state) => state.ui);

  const [values, handleInputChange] = useForm(LoginState);
  const { email, password } = values;

  const handleLogin = (ev) => {
    ev.preventDefault();
    dispatch(initLoginAction(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLoginAction());
  };

  return (
    <div className="animate__animated animate__fadeIn animate__faster login-container">
      <div className="login-container__header">
        <h3 className="login-container__header__title">Login</h3>
        <label className="login-container__header__subheader">
          Save your daily experiences and track your growth!
        </label>
      </div>
      <button className="auth-google" onClick={() => handleGoogleLogin()}>
        <img
          className="google-icon"
          src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          alt="google button"
        />
        <b>Sign in with Google</b>
      </button>
      <div className="auth-separator">
        <span>or Sign with Email</span>
      </div>
      <form type="submit" onSubmit={handleLogin}>
        <div className="headers">
          <div className="textfield-container">
            <label>
              Email<b> *</b>
            </label>
            <input
              type="email"
              className="input-container"
              placeholder="Email"
              name="email"
              autoComplete="new-password"
              value={email}
              onChange={handleInputChange}
            />
          </div>
          <div className="textfield-container">
            <label>
              Password<b> *</b>
            </label>
            <input
              type="password"
              className="input-container"
              placeholder="Password"
              name="password"
              value={password}
              onChange={handleInputChange}
            />
          </div>
          <button
            type="submit"
            className="auth-btn pointer"
            disabled={loading}
            onClick={(e) => handleLogin(e)}
          >
            <b>Login</b>
          </button>
        </div>
        <div className="footers">
          <label className="paragraph">
            <span>Not registered yet?</span>
            <Link to="/auth/register" className="link">
              Create an Account
            </Link>
          </label>
        </div>
      </form>
    </div>
  );
};
