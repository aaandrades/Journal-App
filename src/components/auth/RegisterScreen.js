import React from "react";
import { Link } from "react-router-dom";
import { useForm } from "../../Hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import { RegisterState } from "../../states/InitialStates";
import {
  setErrorAction,
  setRemoveErrorAction,
} from "../../Redux/Actions/uiActions";
import { initRegisterAction } from "../../Redux/Actions/authActions";

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector((state) => state.ui);
  const [values, handleInputChange] = useForm(RegisterState);
  const { name, email, password, password2 } = values;

  const handleRegister = (event) => {
    event.preventDefault();
    if (validateForm()) {
      dispatch(initRegisterAction(name, email, password));
    }
  };

  const validateForm = () => {
    const emailRegex =
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/g;

    if (name.trim().length === 0) {
      dispatch(setErrorAction("Name must be required"));
      return false;
    } else if (!emailRegex.test(email)) {
      dispatch(setErrorAction("Email its incorrect"));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(
        setErrorAction(
          "Password should be at least 6 characters and match each other"
        )
      );
      return false;
    }
    dispatch(setRemoveErrorAction());
    return true;
  };

  return (
    <div className="animate__animated animate__fadeIn animate__faster">
      <h3 className="mb-5">Register</h3>
      <form type="submit" onSubmit={handleRegister}>
        <div className="headers">
          {msgError && <div className="auth__alert-error mb-5">{msgError}</div>}
          <input
            type="text"
            value={values.name}
            className="input-container"
            placeholder="Name"
            name="name"
            autoComplete="new-password"
            onChange={handleInputChange}
          />
          <input
            type="email"
            value={values.email}
            className="input-container"
            placeholder="Email"
            name="email"
            autoComplete="new-password"
            onChange={handleInputChange}
          />
          <input
            type="password"
            value={values.password}
            className="input-container"
            placeholder="Password"
            name="password"
            autoComplete="off"
            onChange={handleInputChange}
          />
          <input
            type="password"
            value={values.password2}
            className="input-container"
            placeholder="Confirm password"
            name="password2"
            onChange={handleInputChange}
          />
          <button
            type="submit"
            className="button pointer"
            onClick={handleRegister}
          >
            Register
          </button>
        </div>
        <div className="footers">
          <span className="paragraph">
            <Link to="/auth/login" className="link">
              Already registered?
            </Link>
          </span>
        </div>
      </form>
    </div>
  );
};
