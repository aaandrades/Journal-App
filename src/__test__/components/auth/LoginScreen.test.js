/**
 * @jest-environment node
 */
import React from "react";
import { mount } from "enzyme";
import "jsdom-global/register";
import "@testing-library/jest-dom";
import { LoginScreen } from "../../../components/auth/LoginScreen";
import { Provider } from "react-redux";

import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import {
  initLoginAction,
  startGoogleLoginAction,
} from "../../../Redux/Actions/authActions";

jest.mock("../../../Redux/Actions/authActions", () => ({
  startGoogleLoginAction: jest.fn(),
  initLoginAction: jest.fn(),
}));

const initialStore = {
  auth: { uid: "Testing", name: "Andyman" },
  ui: { loading: false, msgError: null },
};

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

let store = mockStore(initialStore);
store.dispatch = jest.fn();

describe("LoginScreen - Component", () => {
  beforeEach(() => {
    store = mockStore(initialStore);
    jest.clearAllMocks();
  });

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <LoginScreen />
      </MemoryRouter>
    </Provider>
  );

  test("should match to snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispath login action with google", () => {
    wrapper.find(".google-btn").prop("onClick")();
    expect(startGoogleLoginAction).toHaveBeenCalled();
  });

  test("should dispath login with email/password", () => {
    wrapper.find("form").prop("onSubmit")({ preventDefault() {} });
    expect(initLoginAction).toHaveBeenCalled();
  });
});
