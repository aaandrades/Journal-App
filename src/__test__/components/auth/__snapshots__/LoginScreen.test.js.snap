// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen - Component should match to snapshot 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster login-container"
>
  <div
    className="login-container__header"
  >
    <h3
      className="login-container__header__title"
    >
      Login
    </h3>
    <label
      className="login-container__header__subheader"
    >
      Save your daily experiences and track your growth!
    </label>
  </div>
  <button
    className="auth-google"
    onClick={[Function]}
  >
    <img
      alt="google button"
      className="google-icon"
      src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
    />
    <b>
      Sign in with Google
    </b>
  </button>
  <div
    className="auth-separator"
  >
    <span>
      or Sign with Email
    </span>
  </div>
  <form
    onSubmit={[Function]}
    type="submit"
  >
    <div
      className="headers"
    >
      <div
        className="textfield-container"
      >
        <label>
          Email
          <b>
             *
          </b>
        </label>
        <input
          autoComplete="new-password"
          className="input-container"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value="testaccount@test.com"
        />
      </div>
      <div
        className="textfield-container"
      >
        <label>
          Password
          <b>
             *
          </b>
        </label>
        <input
          className="input-container"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value="123123"
        />
      </div>
      <button
        className="auth-btn pointer"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        <b>
          Login
        </b>
      </button>
    </div>
    <div
      className="footers"
    >
      <label
        className="paragraph"
      >
        <span>
          Not registered yet?
        </span>
        <a
          className="link"
          href="/auth/register"
          onClick={[Function]}
        >
          Create an Account
        </a>
      </label>
    </div>
  </form>
</div>
`;
